connect_four.py

```diff
@@ -205,20 +239,42 @@ def play_game(self):
         self.__game_over()
 
 
-new_game = GameLoop()
+'''new_game = GameLoop()
 
 while new_game.play:
-    new_game.play_game()
+    new_game.play_game()''' #UNCOMMENT TO PLAY THE GAME
 
 
-def tests() -> bool:
-    return None
-    #TODO add tests ..............................
+def tests() -> None:
     #check if board is printed correctly
     #check if pieces are added correctly
     #check user input (char vs string for column)
     #check if "column is full" check works
     #check if "column doesn't exist" works
-    #check if win is correctly detected vertically, horizontally, diagonally
+    #check if win is correctly detected vertically
+    new_game = GameLoop(False, ['A', 'B', 'A', 'B', 'A', 'B', 'A', 'B','A', 'B', 'A', 'B',])
+    while new_game.play:
+        new_game.play_game()
+    assert new_game.winner == "player1", "Player1 should have won, but did not."
```

Comment:
I feel like you could make the code a little bit more "functional" here in that you could have the result of new_game.play_game() return a player, rather than running "play_game", then checking the state of winner.

The reason for this is that the additional state of winner is a little tricky to reason about. There could be bugs if you check the state of winner at the wrong time (before the game begins). I think a more intuitive way to think about this is there is a result that occurs as a result of the "play_game" method executing. The alternative chosen here is mutating the values of a game class which, while is technically "object oriented", I lean towards the functional programming paradigm in this instance. 

Just as an aside, the OOP vs FP debate is going to rage on but I think even proponents of OOP have recognized the benefits of immutability and pure function. I'm more of an FP camp type of person but I do feel as a whole the industry is trending more in that direction. This isn't to say that I think FP is better in all cases- rather, let's take the good from both language paradigms, and one of the great things about FP is immutability, so let's try to abide by that whenever possible.

This could be confusing, let me know if this makes sense